types/index.d.ts

```diff
@@ -0,0 +1,166 @@
+// Minimum TypeScript Version: 4.2
+
+// NOTE: `dtslint` needs 4.2 even though these definitions should work with 4.1.
+
+declare module "hyperapp" {
+  // `app()` initiates a Hyperapp application. `app()` along with runners and
+  // subscribers are the only places where side effects are allowed.
+  function app<S>(props: App<S>): void
+
+  // `h()` builds a virtual DOM node.
+  function h<S, T extends string = string>(
+    // Tags cannot be empty strings.
+    tag: T extends "" ? never : T,
+    props: PropList<S>,
+    children?: VNode<S> | readonly VNode<S>[]
+  ): VDOM<S>
+
+  // `memo()` stores a view along with data for it.
+  function memo<
+    S,
+    D extends string | any[] | Record<string, any>
+  >(
+    view: View<D>,
+    data: D
+  ): VDOM<S>
+
+  // `text()` creates a virtual DOM node representing plain text.
+  function text<T, S>(
+    value: T extends (symbol | ((..._: any[]) => any)) ? never : T,
+    node?: Node
+  ): VDOM<S>
+
+  // ---------------------------------------------------------------------------
+
+  // A Hyperapp application instance has an initial state and a base view.
+  // It must also be mounted over an available DOM element.
+  type App<S> = Readonly<{
+    init: StateFormat<S> | Action<S>
+    view: View<S>
+    node: Node
+    subscriptions?: Subscriptions<S>
+    middleware?: Middleware<S>
+  }>
+
+  // A view builds a virtual DOM node representation of the application state.
+  type View<S> = (state: State<S>) => VDOM<S>
+
+  // The subscriptions function manages a set of subscriptions.
+  type Subscriptions<S> = (state: State<S>) => Subscription<S>[]
+
+  // A subscription represents subscriber activity.
+  type Subscription<S, D = any> = boolean | undefined | SubscriberDescriptor<S, D> | Unsubscribe
+
+  // A subscriber reacts to subscription updates.
+  type SubscriberDescriptor<S, D> = [Subscriber<S, D>, Payload<D>]
```

Comment:
Do we need this `SubscriberDescriptor ` type? Why not just inline `[Subscriber<S, D>, Payload<D>]` in the `Subscription` definition?

Reply:
Got this in the latest changes.